### PR TITLE
IGMP processing

### PR DESCRIPTION
**Pull request by astro**
This is on top of #177 

Once you start review, please keep in mind that this requires a **rust-managed** release with the [ManagedMap.iter()](https://github.com/m-labs/rust-managed/pull/6) pull request.

### CODE REVIEW

File: src/iface/ethernet.rs [hunk 1]
```diff
@@ -55,27 +66,40 @@ pub struct Interface<'b, 'c, DeviceT: for<'d> Device<'d>> {
...
 /// exclusively). However, it is still possible to call methods on its `inner` field.
-struct InterfaceInner<'b, 'c> {
+struct InterfaceInner<'b, 'c, 'e> {
     neighbor_cache:         NeighborCache<'b>,
     ethernet_addr:          EthernetAddress,
     ip_addrs:               ManagedSlice<'c, IpCidr>,
     #[cfg(feature = "proto-ipv4")]
     ipv4_gateway:           Option<Ipv4Address>,
+    #[cfg(feature = "proto-ipv4")]
+    ipv4_mcast_groups:      ManagedMap<'e, Ipv4Address, ()>,
```
Comment (dlrobertson) [inline src/iface/ethernet.rs]: multicast addresses are used in IPv6 as well. See: [MLDv2]. The first implementation may only store `IpAddress::Ipv4` variants in this map, but I think the map should contain `IpAddress` instead of `Ipv4Address`.

[MLDv2]: https://tools.ietf.org/html/rfc3810

File: src/iface/ethernet.rs [hunk 2]
```diff
@@ -159,9 +187,22 @@ impl<'b, 'c, DeviceT> InterfaceBuilder<'b, 'c, DeviceT>
         self
     }
 
+    /// Provide storage for multicast groups
+    ///
+    /// Add them to a finalized `Interface` using [join_multicast_group].
+    ///
+    /// [join_multicast_group]: struct.EthernetInterface.html#method.join_multicast_group
+    #[cfg(feature = "proto-ipv4")]
+    pub fn ipv4_mcast_groups<T>(mut self, ipv4_mcast_groups: T) -> InterfaceBuilder<'b, 'c, 'e, DeviceT>
```
Comment (dlrobertson) [inline src/iface/ethernet.rs]: Again, I don't think this should be IPv4 specific.

File: src/iface/ethernet.rs [hunk 3]
```diff
@@ -262,6 +323,70 @@ impl<'b, 'c, DeviceT> Interface<'b, 'c, DeviceT>
...
+
+                if let Some(pkt) = self.inner.igmp_report_packet(IgmpVersion::Version2, addr) {
+                    // Send initial membership report
+                    let tx_token = self.device.transmit().ok_or(Error::Exhausted)?;
+                    self.inner.dispatch(tx_token, timestamp, pkt)?;
+                }
+
+                Ok(())
+            }
+            // Multicast is not yet implemented for other address families
```
Comment (dlrobertson) [inline src/iface/ethernet.rs]: I think you mean `IGMP` is not implemented in other address families. Multicast is a requirement for IPv6. See [MLDv2].

[MLDv2]: https://tools.ietf.org/html/rfc3810

File: src/iface/ethernet.rs [hunk 4]
```diff
@@ -262,6 +323,70 @@ impl<'b, 'c, DeviceT> Interface<'b, 'c, DeviceT>
...
+
+                if let Some(pkt) = self.inner.igmp_leave_packet(addr) {
+                    // Send group leave packet
+                    let tx_token = self.device.transmit().ok_or(Error::Exhausted)?;
+                    self.inner.dispatch(tx_token, timestamp, pkt)?;
+                }
+
+                Ok(())
+            }
+            // Multicast is not yet implemented for other address families
```
Comment (dlrobertson) [inline src/iface/ethernet.rs]: Same.

File: src/iface/ethernet.rs [hunk 5]
```diff
@@ -514,16 +703,42 @@ impl<'b, 'c> InterfaceInner<'b, 'c> {
...
+            .filter_map(
+                |addr| match addr {
+                    &IpCidr::Ipv4(cidr) => Some(cidr.address()),
+                    _ => None,
+                })
+            .next()
+    }
+
+    /// Check whether the interface listens to given destination multicast IP address.
+    #[cfg(feature = "proto-ipv4")]
```
Comment (dlrobertson) [inline src/iface/ethernet.rs]: Again, shouldn't be IPv4 specific.

File: src/iface/ethernet.rs (hunk 1 above)
Comment (astro) [reply src/iface/ethernet.rs]: Unlike `ip_addrs` (for IPv4) you may join a number of of groups. I didn't want to waste 16 bytes per IPv4 address. My use-case is Embedded Rust.

File: src/iface/ethernet.rs (hunk 3 above)
Comment (astro) [reply src/iface/ethernet.rs]: MLD is not yet implemented in smoltcp.

File: src/iface/ethernet.rs (hunk 1 above)
Comment (dlrobertson) [reply src/iface/ethernet.rs]: Yes, it may optimize space for use-cases that are IPv4 only, but for use cases that are IPv4 and IPv6 we then have two `ManagedMap`s one for IPv4 and the other for IPv6.

File: src/iface/ethernet.rs (hunk 3 above)
Comment (dlrobertson) [reply src/iface/ethernet.rs]: Ah, this comment made me think it was saying Multicast doesn't exist for other families. My bad.

File: src/iface/ethernet.rs (hunk 1 above)
Comment (astro) [reply src/iface/ethernet.rs]: Even in dual-stack mode we save 12 bytes per IPv4 multicast group. Do you think that is not worth it?

Comment (astro) [reply src/iface/ethernet.rs]: I would really like to keep IPv4 and IPv6 mcast groups in separate maps.

Memory space efficiency shouldn't depend on disabling IPv6. Seen that way, doing that could even reduce IPv6 adoption.

Comment (dlrobertson) [reply src/iface/ethernet.rs]: I would still prefer that we have one map, but I can be okay with separate maps and I don't think it should keep this PR from getting merged.

**Comment by m-labs-homu** [comment]
:umbrella: The latest upstream changes (presumably 53abe8c5202b6dda714a0f38e33faf3a49e1191d) made this pull request unmergeable. Please resolve the merge conflicts.

**Comment by astro** [comment]
I've rebased onto current master.

File: src/iface/ethernet.rs [hunk 6]
```diff
@@ -38,14 +44,19 @@ use socket::UdpSocket;
 use socket::TcpSocket;
 use super::{NeighborCache, NeighborAnswer};
 
+#[cfg(feature = "proto-ipv4")]
+const IPV4_MCAST_ALL_SYSTEMS: [u8; 4] = [224, 0, 0, 1];
```
Comment (whitequark) [inline src/iface/ethernet.rs]: These should be associated consts in `Ipv4Address`.

File: src/iface/ethernet.rs [hunk 7]
```diff
@@ -262,6 +323,70 @@ impl<'b, 'c, DeviceT> Interface<'b, 'c, DeviceT>
         InterfaceInner::check_ethernet_addr(&self.inner.ethernet_addr);
     }
 
+    /// Add an address to a list of subscribed multicast IP addresses
+    ///
+    /// Returns whether address was added successfully, and, if
+    /// possible, whether an initial immediate announcement has been
+    /// sent.
+    // timestamp is unused in absence of feature "proto-ipv4"
```
Comment (whitequark) [inline src/iface/ethernet.rs]: Just call it `_timestamp`, this is preferred over `#[allow(unused)]`.

File: src/iface/ethernet.rs [hunk 8]
```diff
@@ -339,7 +470,12 @@ impl<'b, 'c, DeviceT> Interface<'b, 'c, DeviceT>
...
             let processed_any = self.socket_ingress(sockets, timestamp)?;
             let emitted_any   = self.socket_egress(sockets, timestamp)?;
-            if processed_any || emitted_any {
+            #[cfg(feature = "proto-ipv4")]
+            let reported_any  = self.igmp_egress(timestamp)?;
+            #[cfg(not(feature = "proto-ipv4"))]
+            let reported_any  = false;
+
+            if processed_any || emitted_any || reported_any {
                 readiness_may_have_changed = true;
```
Comment (whitequark) [inline src/iface/ethernet.rs]: AFAICT emitting an IGMP packet cannot change readiness of any socket.

File: src/iface/ethernet.rs [hunk 9]
```diff
@@ -514,16 +703,41 @@ impl<'b, 'c> InterfaceInner<'b, 'c> {
...
+    }
+
     fn process_ethernet<'frame, T: AsRef<[u8]>>
                        (&mut self, sockets: &mut SocketSet, timestamp: Instant, frame: &'frame T) ->
                        Result<Packet<'frame>>
     {
         let eth_frame = EthernetFrame::new_checked(frame)?;
 
-        // Ignore any packets not directed to our hardware address.
+        // Check whether to keep the packet
```
Comment (whitequark) [inline src/iface/ethernet.rs]: Looks like a pointless change (here and below).

File: src/iface/ethernet.rs [hunk 10]
```diff
@@ -734,6 +950,80 @@ impl<'b, 'c> InterfaceInner<'b, 'c> {
         }
     }
 
+    /// Host duties of the **IGMPv2** protocol
+    ///
+    /// and we are not worried about routing i.e. we can ignore
```
Comment (whitequark) [inline src/iface/ethernet.rs]: Lots of missing words, punctuation, bad formatting, etc in this doc comment.

File: src/phy/sys/raw_socket.rs [hunk 11]
```diff
@@ -68,7 +68,7 @@ impl RawSocketDesc {
         unsafe {
             let len = libc::send(self.lower, buffer.as_ptr() as *const libc::c_void,
                                  buffer.len(), 0);
-            if len == -1 { Err(io::Error::last_os_error()).unwrap() }
+            if len == -1 { return Err(io::Error::last_os_error()) }
```
Comment (whitequark) [inline src/phy/sys/raw_socket.rs]: Why this change? The unwrap was deliberate.

File: src/socket/raw.rs [hunk 12]
```diff
@@ -154,6 +154,8 @@ impl<'a, 'b> RawSocket<'a, 'b> {
     pub(crate) fn dispatch<F>(&mut self, checksum_caps: &ChecksumCapabilities, emit: F) ->
                              Result<()>
             where F: FnOnce((IpRepr, &[u8])) -> Result<()> {
+        // checksum_caps is unused in absence of feature "proto-ipv4"
+        #[allow(unused)]
```
Comment (whitequark) [inline src/socket/raw.rs]: Same as above re: `_timeout`.

File: src/phy/sys/raw_socket.rs (hunk 11 above)
Comment (astro) [reply src/phy/sys/raw_socket.rs]: Could you please elaborate why you should not be able to gracefully handle the error?

Comment (whitequark) [reply src/phy/sys/raw_socket.rs]: IIRC, my reasoning was that this call should never fail.

**Comment by astro** [comment]
Ready for the next round of review :-)

File: src/iface/ethernet.rs [hunk 13]
```diff
@@ -484,9 +610,64 @@ impl<'b, 'c, DeviceT> Interface<'b, 'c, DeviceT>
...
+                        .iter()
+                        .map(|(group_addr, &())| group_addr)
+                        .skip_while(|group_addr| *group_addr != &addr);
+                    if next_addrs.next() != Some(&addr) {
+                        // Current multicast group has been left, don't know where to continue
+                        self.inner.igmp_report_state = IgmpReportState::Inactive;
+                        return Ok(false)
+                    }
+                    next_addrs.next()
+                        .map(|next_addr| next_addr.clone())
```
Comment (whitequark) [inline src/iface/ethernet.rs]: Ipv4Address is Copy.

File: src/iface/ethernet.rs [hunk 14]
```diff
@@ -484,9 +610,64 @@ impl<'b, 'c, DeviceT> Interface<'b, 'c, DeviceT>
...
+
+                self.inner.igmp_report_state = IgmpReportState::Inactive;
+                Ok(true)
+            }
+            IgmpReportState::ToGeneralQuery(version, timeout, interval, addr) if timestamp >= timeout => {
+                let next_addr = {
+                    let mut next_addrs = self.inner.ipv4_mcast_groups
+                        .iter()
+                        .map(|(group_addr, &())| group_addr)
+                        .skip_while(|group_addr| *group_addr != &addr);
```
Comment (whitequark) [inline src/iface/ethernet.rs]: This is quadratic isn't it? Why not store an index into `ipv4_mcast_groups` and invalidate it when changing the map?

File: src/iface/ethernet.rs [hunk 15]
```diff
@@ -514,16 +695,42 @@ impl<'b, 'c> InterfaceInner<'b, 'c> {
...
+            .filter_map(
+                |addr| match addr {
+                    &IpCidr::Ipv4(cidr) => Some(cidr.address()),
+                    _ => None,
+                })
+            .next()
+    }
+
+    /// Check whether the interface listens to given destination multicast IP address.
+    pub fn has_mcast_group<T: Into<IpAddress>>(&self, addr: T) -> bool {
```
Comment (whitequark) [inline src/iface/ethernet.rs]: This should be conditional on `feature = "proto-igmp"`, like the rest.

Comment (whitequark) [reply src/iface/ethernet.rs]: I also think this function should do an early return if the address being considered is not a multicast address. This will save us the map traversal.

File: src/iface/ethernet.rs [hunk 16]
```diff
@@ -734,6 +944,61 @@ impl<'b, 'c> InterfaceInner<'b, 'c> {
...
+        match igmp_repr {
+            IgmpRepr::MembershipQuery { group_addr, version, max_resp_duration } => {
+                // General Query
+                if group_addr.is_unspecified() && ipv4_repr.dst_addr == Ipv4Address::MCAST_ALL_SYSTEMS {
+                    if let Some((first_group_addr, &())) = self.ipv4_mcast_groups.iter().next() {
+                        let interval = match version {
+                            IgmpVersion::Version1 =>
+                                Duration::from_millis(100),
+                            IgmpVersion::Version2 => {
+                                // No dependence on a random generator
```
Comment (whitequark) [inline src/iface/ethernet.rs]: There should probably be an issue about adding RNG to IGMP code.

File: src/iface/ethernet.rs [hunk 17]
```diff
@@ -734,6 +944,61 @@ impl<'b, 'c> InterfaceInner<'b, 'c> {
...
+                                let intervals = self.ipv4_mcast_groups.len() as u32 + 1;
+                                max_resp_duration / intervals
+                            }
+                        };
+                        self.igmp_report_state = IgmpReportState::ToGeneralQuery(
+                            version,
+                            timestamp + interval,
+                            interval,
+                            first_group_addr.clone()
+                        );
```
Comment (whitequark) [inline src/iface/ethernet.rs]: Can you explain why you cannot just report immediately? Preferably in the comments.

File: src/iface/ethernet.rs [hunk 18]
```diff
@@ -1174,16 +1445,54 @@ impl<'b, 'c> InterfaceInner<'b, 'c> {
             f(ip_repr, payload)
         })
     }
+
+    #[cfg(all(feature = "proto-ipv4", feature = "proto-igmp"))]
+    fn igmp_report_packet<'any>(&self, version: IgmpVersion, group_addr: Ipv4Address) -> Option<Packet<'any>> {
+        self.ipv4_address().map(|iface_addr| {
```
Comment (whitequark) [inline src/iface/ethernet.rs]: I would prefer a `let iface_addr = match self.ipv4_address() { ... }` here.

File: src/socket/raw.rs [hunk 19]
```diff
@@ -155,13 +155,13 @@ impl<'a, 'b> RawSocket<'a, 'b> {
                              Result<()>
             where F: FnOnce((IpRepr, &[u8])) -> Result<()> {
         fn prepare<'a>(protocol: IpProtocol, buffer: &'a mut [u8],
-                   checksum_caps: &ChecksumCapabilities) -> Result<(IpRepr, &'a [u8])> {
+                   _checksum_caps: &ChecksumCapabilities) -> Result<(IpRepr, &'a [u8])> {
```
Comment (whitequark) [inline src/socket/raw.rs]: Why?

File: src/wire/igmp.rs [hunk 20]
```diff
@@ -173,7 +173,7 @@ impl<T: AsRef<[u8]> + AsMut<[u8]>> Packet<T> {
 #[derive(Debug, PartialEq, Eq, Clone)]
 pub enum Repr {
     MembershipQuery {
-        max_resp_time: Duration,
+        max_resp_duration: Duration,
```
Comment (whitequark) [inline src/wire/igmp.rs]: Rebase error? I specifically renamed it to `max_resp_time`, since that's how the field is called in the RFC.

File: src/iface/ethernet.rs [hunk 21]
```diff
@@ -1174,16 +1445,54 @@ impl<'b, 'c> InterfaceInner<'b, 'c> {
...
+                group_addr,
+                version,
+            };
+            let pkt = Packet::Igmp((Ipv4Repr {
+                src_addr:    iface_addr.clone(),
+                // Send to the group being reported
+                dst_addr:    group_addr,
+                protocol:    IpProtocol::Igmp,
+                payload_len: igmp_repr.buffer_len(),
+                hop_limit:   1,
```
Comment (whitequark) [inline src/iface/ethernet.rs]: From #60:
> the TTL value for group query has to be set to 1, and to 255 for membership report (according to Wireshark and RFC 3717)

File: examples/multicast.rs [hunk 22]
```diff
@@ -0,0 +1,111 @@
+#[macro_use]
```
[the remark this answers is not in the document]
Comment (astro) [reply examples/multicast.rs]: @podhrmic did that

File: src/wire/igmp.rs (hunk 20 above)
Comment (astro) [reply src/wire/igmp.rs]: I thought otherwise, because in `Repr` we're no longer dealing with the `max_resp_time` code but with a proper `Duration`. I'll rename it back if you don't agree that that's a significant difference between the two.

File: src/socket/raw.rs (hunk 19 above)
Comment (astro) [reply src/socket/raw.rs]: Unused when `feature = "proto-ipv4"` is **absent**.

File: src/iface/ethernet.rs (hunk 21 above)
Comment (astro) [reply src/iface/ethernet.rs]: I don't like the semantics of TTL=255, that is: reporting group membership to every group listener across routers. I'm not seeing it in the wild either.

As this is IGMPv2 being implemented here, not [IP over Optical Networks: A Framework](https://tools.ietf.org/html/rfc3717)(?), from [RFC 2236](https://tools.ietf.org/html/rfc2236):
> All IGMP messages described in this document are sent with IP TTL 1, and contain the IP Router Alert option [RFC 2113] in their IP header.
> […]
> When a group's timer expires, the host multicasts a Version 2 Membership Report to the group, with IP TTL of 1.

Ok, I didn't know about the Router Alert option… but I guess the comment regarding TTL might have been a mistake.

File: src/wire/igmp.rs (hunk 20 above)
Comment (whitequark) [reply src/wire/igmp.rs]: The *concept* is the same (maximum response time), the *units* are different (floating point in i8 vs Duration). You could call it `max_resp_time_fp` and `max_resp_time_duration` but that's annoyingly long.

File: src/iface/ethernet.rs [hunk 23]
```diff
@@ -55,27 +61,40 @@ pub struct Interface<'b, 'c, DeviceT: for<'d> Device<'d>> {
...
 /// methods on the `Interface` in this time (since its `device` field is borrowed
 /// exclusively). However, it is still possible to call methods on its `inner` field.
-struct InterfaceInner<'b, 'c> {
+struct InterfaceInner<'b, 'c, 'e> {
     neighbor_cache:         NeighborCache<'b>,
     ethernet_addr:          EthernetAddress,
     ip_addrs:               ManagedSlice<'c, IpCidr>,
     #[cfg(feature = "proto-ipv4")]
     ipv4_gateway:           Option<Ipv4Address>,
+    #[cfg(feature = "proto-ipv4")]
```
[the remark this answers is not in the document]
Comment (astro) [reply src/iface/ethernet.rs]: Right! Thank you for pointing out.

File: src/wire/igmp.rs (hunk 20 above)
Comment (astro) [reply src/wire/igmp.rs]: Ok!

**Comment by podhrmic** [comment]
What is the status here? @astro do you need help implementing the changes?

**Comment by whitequark** [comment]
It's blocked on review that I haven't had time to perform yet, sorry.

File: src/iface/ethernet.rs (hunk 14 above)
Comment (astro) [reply src/iface/ethernet.rs]: If by index you mean an iterator: cannot borrow from `ipv4_mcast_groups` across multiple calls… any help is appreciated.

**Comment by podhrmic** [comment]
@whitequark can I help here in any way to get this merged?

**Comment by whitequark** [comment]
I'll try to review it soon

**Comment by podhrmic** [comment]
@whitequark any updates here?

**Comment by astro** [comment]
Added commits addressing your latest review. Please verify.

File: examples/multicast.rs [hunk 24]
```diff
@@ -0,0 +1,111 @@
...
+        }
+        {
+            let mut socket = sockets.get::<UdpSocket>(udp_handle);
+            if !socket.is_open() {
+                socket.bind(MDNS_PORT).unwrap()
+            }
+
+            if socket.can_recv() {
+                // For display purposes only - normally we wouldn't process incoming IGMP packets
+                // in the applicaiton layer
```
[the remark this answers is not in the document]
Comment (astro) [reply examples/multicast.rs]: It does. I verified.

File: src/iface/ethernet.rs [hunk 25]
```diff
@@ -951,9 +951,15 @@ impl<'b, 'c, 'e> InterfaceInner<'b, 'c, 'e> {
...
-        let igmp_repr = IgmpRepr::parse(&igmp_packet)?;
+                           Result<Packet<'frame>> {
+        let igmp_packet = match IgmpPacket::new_checked(ip_payload) {
+            Ok(igmp_packet) => igmp_packet,
+            Err(_) => return Ok(Packet::None),
+        };
+        let igmp_repr = match IgmpRepr::parse(&igmp_packet) {
+            Ok(igmp_repr) => igmp_repr,
+            Err(_) => return Ok(Packet::None),
+        };
```
Comment (whitequark) [inline src/iface/ethernet.rs]: Why did you do this? smoltcp intentionally pops up errors to the very top of the stack, such that if you receive malformed packets and nothing works, you do not have to wonder as to what is the cause.

Comment (astro) [reply src/iface/ethernet.rs]: You said you want the example to panic on Err. That was happening to me all the time due to IGMPv3 traffic and invalid checksums.

Comment (whitequark) [reply src/iface/ethernet.rs]: Thanks for the explanation. I will revert this part, and make sure examples log errors instead of crashing on them. I did not realize that IGMPv3 traffic would cause this.

**Comment by m-labs-homu** [comment]
:umbrella: The latest upstream changes (presumably 06d18130cd8b25edbcad207bbdbb157f1c1203df) made this pull request unmergeable. Please resolve the merge conflicts.

**Comment by astro** [comment]
I'm verry sorry, but could you please tell me at which point we are in the review process?

**Comment by whitequark** [comment]
@astro To expedite merging, I've fixed all remaining issues I found. Please review them, and squash the commits once you find the state of the branch satisfactory. I'll merge afterwards.

**Comment by astro** [comment]
I corrected the description of IGMP report intervals to general queries in `README.md`.

All your code changes make sense to me. Thanks for the great work!

**Comment by whitequark** [comment]
Thanks for the fixes, and I apologize again for the delay. Can you squash?

**Comment by whitequark** [comment]
@m-labs-homu r+

**Comment by m-labs-homu** [comment]
:pushpin: Commit 777d4bd has been approved by `whitequark`

**Comment by m-labs-homu** [comment]
:hourglass: Testing commit 777d4bd3c419c4fe56ba7f6b8d284afd8ab8e78a with merge bada3f74bd77af2f541de815b6c377b999d16d4d...

**Comment by m-labs-homu** [comment]
:boom: Test timed out

**Comment by whitequark** [comment]
@m-labs-homu retry

**Comment by m-labs-homu** [comment]
:hourglass: Testing commit 777d4bd3c419c4fe56ba7f6b8d284afd8ab8e78a with merge 58a54730dba398aaa5ea88e6c6f89e0234f89f42...

**Comment by m-labs-homu** [comment]
:sunny: Test successful - [status-travis](https://travis-ci.org/m-labs/smoltcp/builds/410611707?utm_source=github_status&utm_medium=notification)
Approved by: whitequark
Pushing 58a54730dba398aaa5ea88e6c6f89e0234f89f42 to master...
